pop/service/service.go

```diff
@@ -0,0 +1,258 @@
+package service
+
+/*
+Service for a Proof-of-Personhood party
+
+Proof-of-personhood parties provide a number of "attendees" with an "anonymous
+token" that enables them to "authenticate" to a service as being part of the
+party.
+
+These parties are held by a number of "organisers" who set up a party by
+defining place, time and purpose of that party and by publishing a
+"party configuration" that is signed by the organisers "conodes".
+At the party, they "register" all attendees' public keys.
+Once the party is over, they create a "party transcript" that is signed by all
+organisers' conodes.
+
+The attendees create their "pop token" by joining their private key to the
+party transcript. They can now use that token to sign a "message" in a "context"
+from a service and send the resulting "signature" and "tag" back to the service.
+
+On the service's side, it can use the party transcript to verify that the
+signature has been created using a private key present in the party transcript.
+The tag will be unique to that attendee/context pair, but another service using
+another context will not be able to link two tags to the same or different
+attendee.
+*/
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"gopkg.in/dedis/crypto.v0/abstract"
+	"gopkg.in/dedis/onet.v1"
+	"gopkg.in/dedis/onet.v1/log"
+	"gopkg.in/dedis/onet.v1/network"
+)
+
+// Name is the name to refer to the Template service from another
+// package.
+const Name = "PoPServer"
+const cfgName = "pop.bin"
+
+var checkConfigID network.MessageTypeID
+var checkConfigReplyID network.MessageTypeID
+
+func init() {
+	onet.RegisterNewService(Name, newService)
+	network.RegisterMessage(&saveData{})
+	checkConfigID = network.RegisterMessage(CheckConfig{})
+	checkConfigReplyID = network.RegisterMessage(CheckConfigReply{})
+}
+
+// Service represents data needed for one pop-party.
+type Service struct {
+	// We need to embed the ServiceProcessor, so that incoming messages
+	// are correctly handled.
+	*onet.ServiceProcessor
+	path string
+	data *saveData
+	// channel to return the configreply
+	ccChannel chan *CheckConfigReply
+}
+
+type saveData struct {
+	// Pin holds the randomly chosen pin
+	Pin string
+	// Public key of linked pop
+	Public abstract.Point
+	// The final statement
+	Final *FinalStatement
+}
+
+// PinRequest prints out a pin if none is given, else it verifies it has the
+// correct pin, and if so, it stores the public key as reference.
+func (s *Service) PinRequest(req *PinRequest) (network.Message, onet.ClientError) {
+	if req.Pin == "" {
```

Comment:
As we can get the server to provide it's signature to a party, we want to make sure that only legit users can do so. So we use the PIN to link the user to the server. The user has to proof he owns the server and has access to it.
This way the services authenticating the users _can_ trust the servers, instead of having to trust only a list of party-ids.